plugins/inventory/kubevirt.py

```diff
@@ -375,61 +387,49 @@ def parse(self, inventory: Any, loader: Any, path: str, cache: bool = True) -> N
         """
         parse is the main entry point of the inventory.
         It checks for availability of the Kubernetes Python client,
-        gets the configuration and runs fetch_objects or
-        if there is a cache it is used instead.
+        gets the configuration, retrieves the cache or runs fetch_objects and
+        populates the inventory.
         """
-        super().parse(inventory, loader, path)
-        cache_key = self._get_cache_prefix(path)
-        config_data = self._read_config_data(path)
-
         if not HAS_K8S_MODULE_HELPER:
             raise KubeVirtInventoryException(
                 "This module requires the Kubernetes Python client. "
                 + f"Try `pip install kubernetes`. Detail: {K8S_IMPORT_EXCEPTION}"
             )
 
-        source_data = None
-        if cache and cache_key in self._cache:
-            try:
-                source_data = self._cache[cache_key]
-            except KeyError:
-                pass
-
-        if not source_data:
-            self.fetch_objects(config_data)
+        super().parse(inventory, loader, path)
 
-    def fetch_objects(self, config_data: Dict) -> None:
-        """
-        fetch_objects populates the inventory with the specified parameters.
-        """
-        if not config_data or not isinstance(config_data, dict):
-            config_data = {}
+        config_data = self._read_config_data(path)
+        cache_key = self.get_cache_key(path)
+        user_cache_setting = self.get_option("cache")
+        attempt_to_read_cache = user_cache_setting and cache
+        cache_needs_update = user_cache_setting and not cache
 
         self.connections_compatibility(config_data)
-        client = get_api_client(**config_data)
-        name = config_data.get(
-            "name", self.get_default_host_name(client.configuration.host)
-        )
-        namespaces = (
-            config_data["namespaces"]
-            if config_data.get("namespaces")
-            else self.get_available_namespaces(client)
-        )
         opts = InventoryOptions(config_data=config_data)
-        if opts.base_domain is None:
-            opts.base_domain = self.get_cluster_domain(client)
-        for namespace in namespaces:
-            self.populate_inventory_from_namespace(client, name, namespace, opts)
+
+        results = {}
+        if attempt_to_read_cache:
+            try:
+                results = self._cache[cache_key]
+            except KeyError:
+                cache_needs_update = True
```

Comment:
I'm not 100% sure, but IMHO, using exceptions for checking if an element is present or not in a data structure is not a good practice. What about:
```python
if attempt_to_read_cache:
     if cache_key in self._cache:
         results = self._cache[cache_key]
     else:
         cache_needs_update = True
```
Using exceptions as flow control, generally speaking, it is not a good idea.

Reply:
results = self._cache.get(cache_key, {})
And checking if the `result` var is `{}` and updating the cache in that case may also work.

Reply:
`self._cache` is not a plain dict and therefore might behave slightly different. This was taken from the Ansible documentation.

https://docs.ansible.com/ansible/9/dev_guide/developing_inventory.html#inventory-cache

Reply:
Ah, I see. Sounds good them.